### PR TITLE
improve some IntSet operations

### PR DESCRIPTION
**Pull request by rfourquet**
* IntSet has a new invariant: the length of the underlying BitArray
  is always a multiple of 64
* the size is set to the minimum possible: the underlying routines
  are in charge of allocating enough (like for other containers);
  this can help to obtain the last element faster (which is useful
  e.g in `rand(::IntSet)`)
* _matched_map avoids extending a BitArray when unecessary
  (but doesn't work anymore for all 2^4 kinds of binary predicate,
  though it could be handled if needed in the future)

A case where this could be slower is for example in:
```
s = IntSet()
for i=1:1000
    push!(s, i)
end
```
But I could not really/reliably measure a difference (disclaimer: `BenchmarkTools` is broken recently on my installation, so I use a hand-made "equivalent"), and `sizehint!` can help here. The improvements for some operations can be huge, e.g.  `symdiff!(a, b)` where `a=IntSet([2^20]), b=IntSet([1])` decreases from 45.07µs to 10ns. Also, something like `rand(s)` where `s=IntSet([1, 2, 3])` decreases from about 980ns to 310ns.

### CODE REVIEW

File: base/intset.jl [hunk 1]
```diff
@@ -159,14 +172,7 @@ function symdiff!(s1::IntSet, s2::IntSet)
...
 
-@inline function in(n::Integer, s::IntSet)
-    if 1 <= n <= length(s.bits)
-        @inbounds b = s.bits[n]
-    else
-        b = false
-    end
-    b
-end
+@inline in(n::Integer, s::IntSet) = get(s.bits, n, false)
```
Comment (rfourquet) [inline base/intset.jl]: Unrelated change, but simpler.

**Review by JeffBezanson — APPROVED** [review]

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23138#issuecomment-320467990) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/3dbb08727df11be632005c24020215ff4473245f/e181e2c_vs_83007fb/report.md). cc @ararslan

**Comment by tkelman** [comment]
stockcorr and array indexing regressions have a chance of being real, should keep an eye on the nightly results

**Comment by fredrikekre** [comment]
Given https://github.com/JuliaLang/julia/commit/ae1719872092b594228a9bfeedfc3fa8e91c9529#commitcomment-23501641 it might be worth looking into the array indexing regression?

**Comment by rfourquet** [comment]
Looking at the very few places in base where `IntSet` are used, I have a hard time understanding how this change could affect these 2 tests... Also, I cannot reproduce locally, i.e. the regressions on those two tests, "stockcorr" and "sumvector_view", stay within 5% (for some reason, the `@benchmark` macro fails in my REPL, but running BaseBenchmarks is ok :-/ )

**Comment by tkelman** [comment]
doesn't inference use them in some key places? otherwise they'd likely be moved out to DataStructures

**Comment by rfourquet** [comment]
Inference uses them, but I thought it wouldn't influence the run time performance and the benchmarks.

**Pull request merged**